README.md
# quickly
The quickly python package will become the successor of python-ly.


ChangeLog
ChangeLog
=========


2022-03-14: quickly-0.7.0

- fix issue #3, allow adding octave to notes in chordmode, but not to the
  inversion
- fix issue #4, relative to absolute yields incorrect results
- removed the quickly.registry module, included languages are now added to
  parce's global registry
- correctly handle \skip with music instead of a single duration (Lily>=2.23.6)
- rhythm.transform() now also handles duration in commands like \tempo, \tuplet,
  \after and \partial
- documentation improvements


2022-02-12: quickly-0.6.0

- requires parce-0.31.0
- small registry module changes to adapt to parce


2022-01-23: quickly-0.5.0

- requires parce-0.30.0
- added time module for time computations
- added key module for key signature and tonic computations
- added rhythm module to manipulate durations in music or a part thereof
- added dom.scope module to find nodes from included files
- modified PitchProcessor:
    - note names now carry a default octave
    - pitch() method now returns a Pitch, old pitch() renamed to process()
- introduced lily.ChordBody and FigureBody, to be able to add Duration at Chord
  and Figure directly
- added lily.Durable as base type for any stuff that can have a Duration
- notes as argument to \relative, \key or \transpose are now Pitch, not Note
- handle single durations in lyricmode, just like Unpitched
- improvements to duration module, add duration.Transform
- added transform() method to all Music nodes, to calculate durations in child
  nodes
- added properties() method to Music node type, to store information during
  processing of descendant music
- added datatypes module for small datatype helper classes, featuring Properties
- documentation improvements


2021-12-31: quickly-0.4.0

- added relative module, with abs2rel and rel2abs functions
- possible to direct generators in Node and Range to not descend in a node
- added node.Range class, to denote a range within a Node tree
- added end parameter to Element.find_descendant(s)
- added Element.find_descendant_right() and -left()
- added dom.edit module to edit a parce document via the DOM node tree
- added pitch.PitchProcessor.find_language and follow_language()
- some methods/functions renamed and API cleanups
- documentation improvements


2021-12-21: quickly-0.3.0

- requires parce-0.28.0
- DOM elegantly handles untransformed parce stuff such as CSS/JS in Html
- Note and positioned Rest now have attrs for octave, accidental and oct_check
- pitch: added Pitch and PitchProcessor to read/write pitch names in all
  languages, with style preferences for writing pitch names
- added lily.Spanner.find_parallel() to find the other end of a spanner
- added transpose module, transpose music in a node or (part of) parce.Document


2021-12-09: quickly-0.2.0

- requires parce-0.27.0
- added registry module to find and guess languages (like parce)
- added LaTeX language and transform (for lilypond-book)
- added element constructor helper for dom.htm module
- added Element.py_dump() to create tree structures in Python code
- simplified dom.read module and API, now that parce's Transform finding
  algorithm is configurable
- small bugfixes, improvements and documentation updates


2021-11-25: quickly-0.1.0

- requires parce-0.26.0
- added html language and transform (for lilypond-book)
- add dom.htm element types (for lilypond-book HTML documents)
- added indent module and support for creating properly indented output
- documentation improvements


2021-11-11: quickly-0.0.9

- requires parce-0.25.0
- Add support for some new LilyPond-2.23 commands
- fixed case where a symbol was interpreted as a pitch eg: \tweak color red do
- fixed some cases where incorrect input was not handled neatly
- documentation improvements


2021-08-08: quickly-0.0.8

- requires parce-0.24.0
- fixed instance dict in Element nodes, added empty __slots__to mixins
- added Node.filter() and Node.instances_of()
- added a logo (two eighth rests, looking like "ly")
- new website at https://quick-ly.info/
- small optimizations and documentation improvements


2021-03-08: quickly-0.0.7

- requires parce-0.23.0
- added quickly.numbering module
- added dom.read module to construct elements/documents from text
- removed some temp methods from dom.util
- Element.copy_with_origin now works properly everywhere
- Node.equals() now works for TextElement, by implementing body_equals()
- improvements to TextElement, optional head value checking
- renamed lily.PropToggle to Toggle
- fixed escape warnings in source files
- added duration.shift_duration()
- moved scm.Int to scm.Number in some forgotten places


2021-02-16: quickly-0.0.6

- required parce version 0.22.0
- lily.Header, lily.Paper and lily.Layout now have easy to use properties to
  set variables like title, composer, etc.
- lily.Document now has version attribute to read/set LilyPond version
- added markup construction helper in quickly.dom.markup
- Scheme transform handles all Scheme number features, e.g. #xb/f (a fraction
  in hex) and even complex numbers


2021-02-11: quickly-0.0.5

- Milestone: quickly correctly parses my Mouvement piece!
  https://github.com/wbsoft/lilymusic/blob/master/berendsen/mouvement.ly
- fixed \rest before duration
- fixed dot missing in scheme pair
- add space after markup commands
- fixed dom.element.build_tree, it lost a node after two \repeat nodes
- parce removed identifier context; we only create Identifier with Assignment
- smartly support \tag and \tweak without direction prepended
- fixed documentation build errors.


2021-02-07: quickly-0.0.4

- still pre-alpha but work in progress
- quickly.dom.lily now has much more elements
- correctly parse lists, property paths etc
- added chordmode and chord modifiers, figuremode
- quickly.lang.lilypond now transforms almost a full document
- commands are combined with their arguments, based on signatures
- work started on documentation with a quickly.dom explanation


2021-01-31: quickly-0.0.3

- still pre-alpha

quickly/time.py
# -*- coding: utf-8 -*-
#
# This file is part of `quickly`, a library for LilyPond and the `.ly` format
#
# Copyright © 2021-2021 by Wilbert Berendsen <info@wilbertberendsen.nl>
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This module is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


"""
Functionality to compute the time length of musical expressions.
"""


import collections
import itertools

from . import datatypes, duration
from .dom import lily


#: The result value (if not None) of the :meth:`~Time.position` and
#: :meth:`~Time.duration` methods.
Result = collections.namedtuple("Result", "node time")
Result.node.__doc__ = "The topmost Music expression."
Result.time.__doc__ = "The position or duration time value."


class Time:
    r"""Compute the length of musical expressions.

    A :class:`~.scope.Scope`, if given using the ``scope`` parameter, is used
    to resolve include files. If no scope is given, only searches the current
    DOM document for variable assignments.

    If a scope is given, include commands are followed and ``wait`` determines
    whether to wait for ongoing transformations of external DOM documents. If
    wait is False, and a transformation is not yet finished the include is not
    followed.

    An example::

        >>> import parce, quickly.time
        >>> d = parce.Document(quickly.find('lilypond'), r'''
        ... music = { c4 d e f }
        ...
        ... { c2 \music g a b8 g f d }
        ... ''', transformer=True)
        >>> m = d.get_transform(True)
        >>> m.dump()
        <lily.Document (2 children)>
         ├╴<lily.Assignment music (3 children)>
         │  ├╴<lily.Identifier (1 child)>
         │  │  ╰╴<lily.Symbol 'music' [1:6]>
         │  ├╴<lily.EqualSign [7:8]>
         │  ╰╴<lily.MusicList (4 children) [9:21]>
         │     ├╴<lily.Note 'c' (1 child) [11:12]>
         │     │  ╰╴<lily.Duration Fraction(1, 4) [12:13]>
         │     ├╴<lily.Note 'd' [14:15]>
         │     ├╴<lily.Note 'e' [16:17]>
         │     ╰╴<lily.Note 'f' [18:19]>
         ╰╴<lily.MusicList (8 children) [23:49]>
            ├╴<lily.Note 'c' (1 child) [25:26]>
            │  ╰╴<lily.Duration Fraction(1, 2) [26:27]>
            ├╴<lily.IdentifierRef 'music' [28:34]>
            ├╴<lily.Note 'g' [35:36]>
            ├╴<lily.Note 'a' [37:38]>
            ├╴<lily.Note 'b' (1 child) [39:40]>
            │  ╰╴<lily.Duration Fraction(1, 8) [40:41]>
            ├╴<lily.Note 'g' [42:43]>
            ├╴<lily.Note 'f' [44:45]>
            ╰╴<lily.Note 'd' [46:47]>
        >>> t=quickly.time.Time()
        >>> t.position(m[1][0])         # first note in second expression
        Result(node=<lily.MusicList (8 children) [23:49]>, time=0)
        >>> t.position(m[1][1])         # \music identifier ref
        Result(node=<lily.MusicList (8 children) [23:49]>, time=Fraction(1, 2))
        >>> t.position(m[1][2])         # the g after the \music ref
        Result(node=<lily.MusicList (8 children) [23:49]>, time=Fraction(3, 2))
        >>> t.length(m[0][2])           # the \music expression
        Fraction(1, 1)
        >>> t.length(m[1])              # total length of second expression
        Fraction(3, 1)                  # referenced \music correctly counted in :)
        >>> t.duration(m[1][0], m[1][2])# length of "c2 \music g" part (g has duration 2)
        Result(node=<lily.MusicList (8 children) [23:49]>, time=Fraction(2, 1))

    There are convenient methods to get the musical position of a parce
    :class:`~parce.Cursor`::

        >>> c = parce.Cursor(d, 44, 47)
        >>> c.text()                    # two notes
        'f d'
        >>> t.cursor_position(c)        # length or music before the cursor
        Result(node=<lily.MusicList (8 children) [23:49]>, time=Fraction(11, 4))
        >>> t.cursor_duration(c)        # duration of the selected music
        Result(node=<lily.MusicList (8 children) [23:49]>, time=Fraction(1, 4))

    LilyPond music functions that alter durations are recognized, and are
    abstracted in simple transformations that alter log, dotcount and/or
    scaling. An example::

        >>> from quickly.dom import read
        >>> m = read.lily(r"\tuplet 3/2 { c8 d e }")
        >>> t.length(m)
        Fraction(1, 4)
        >>> m[1][1]                     # a single note in the tuplet
        <lily.Note 'd'>
        >>> t.length(m[1][1])
        Fraction(1, 12)
        >>> m = read.lily(r"\shiftDurations #1 #1 { c4 d e f }")
        >>> t.length(m)         # note value halved and dot added, so should be 3/4
        Fraction(3, 4)
        >>> m[2][2]
        <lily.Note 'e'>
        >>> t.length(m[2][2])   # autodiscovers the current duration transform
        Fraction(3, 16)

    .. note::

       As a :class:`Time` instance uses some caching for the duration of
       individual notes, don't rely on its computations while also modifying
       durations of music notes, rests etc.

    """
    def __init__(self, scope=None, wait=False):
        self.scope = scope  #: Our :class:`~.scope.Scope`.
        self.wait = wait    #: If True, parce transformations are waited for.
        self.get_duration = lily.duration_getter()

    def __repr__(self):
        return "<{} scope={} wait={}>".format(
            type(self).__name__, self.scope, self.wait)

    @staticmethod
    def _music_child(node):
        """Return the topmost Music child or None."""
        # avoid picking one note of a chord
        if isinstance(node, lily.Chord):
            return node
        for chord in node << lily.Chord:
            return chord
        while node:
            if isinstance(node, lily.Music) or isinstance(node.parent, lily.Music):
                return node
            node = node.parent

    @staticmethod
    def _preceding_music(node):
        """Return a two-tuple(music, trail).

        The ``node`` must be (a child of) a :class:`~.dom.lily.Music` instance.
        The returned music is the topmost Music ancestor and the trail lists
        the indices of the children upto the node.

        """
        trail = []
        for p, i in node.ancestors_with_index():
            if not isinstance(p, lily.Music):
                break
            trail.append(i)
            node = p
        trail.reverse()
        return node, trail

    def length(self, node):
        """Return the musical length of this node."""
        context = TimeContext(self)
        if isinstance(node, (lily.Music, lily.Reference)):
            ancestors = list(itertools.takewhile(lily.is_music, node.ancestors()))
            for p in reversed(ancestors):
                context = context.enter(p)
        return context.length(node)

    def position(self, node, include=False):
        """Return a :class:`Result` two-tuple(node, time).

        The ``node`` argument must be (a child of) a :class:`~.dom.lily.Music`
        instance.

        The returned ``node`` is the topmost Music element, and the ``time`` is
        the computed length of all preceding music nodes. If ``include`` is
        True, the node's length itself is also added.

        """
        music, trail = self._preceding_music(node)
        context = TimeContext(self)
        context, node, length = context._follow_trail(music, trail)
        if include:
            length += context.length(node)
        return Result(music, length)

    def duration(self, start_node, end_node):
        """Return a :class:`Result` two-tuple(node, time) or None.

        The returned ``node`` is the topmost Music element both nodes must be a
        descendant of. Both nodes must be (children of)
        :class:`~.dom.lily.Music` nodes. If they don't share the same ancestor,
        None is returned. The returned ``time`` value can be negative when the
        end node precedes the start node.

        """
        music, start_trail = self._preceding_music(start_node)
        end_music, end_trail = self._preceding_music(end_node)
        if end_music is not music:
            return

        node = music
        context = TimeContext(self)

        # common part, just follow transform
        index = -1
        for index, (pos, end) in enumerate(zip(start_trail, end_trail)):
            context = context.enter(node)
            if pos != end:
                break
            node = node[pos]
        else:
            index += 1

        # compute time only for differing part of the trails
        start_time = context._follow_trail(node, start_trail[index:])[2]
        context, node, end_time = context._follow_trail(node, end_trail[index:])
        end_time += context.length(node)
        return Result(music, end_time - start_time)

    def cursor_position(self, cursor):
        """Return a :class:`Result` two-tuple(node, time) or None.

        The ``node`` is the music expression the cursor is in, and the ``time``
        is the time offset from the start of that expression to the cursor's
        position.

        Returns None if the cursor is not in music.

        """
        dom = cursor.document().get_transform(self.wait)
        if dom:
            node = dom.find_descendant_right(cursor.pos)
            c = self._music_child(node)
            if c:
                return self.position(c)

    def cursor_duration(self, cursor):
        """Return a :class:`Result` two-tuple(node, time) or None.

        The ``node`` is the music expression the cursor is in, and the ``time``
        is the length of the selected music fragment.

        Returns None if there's no selection or the selection's start and/or
        end are not in music, or in different music expressions.

        """
        if cursor.has_selection():
            pos, end = cursor.selection()
            dom = cursor.document().get_transform(self.wait)
            if dom:
                n = dom.find_descendant_right(pos)
                if n:
                    start_node = self._music_child(n)
                    if start_node:
                        n = dom.find_descendant_left(end)
                        if n:
                            end_node = self._music_child(n)
                            if end_node:
                                result = self.duration(start_node, end_node)
                                if result and result.time >= 0:
                                    return result


class TimeContext:
    """Encapsulates the transform and properties during time calculations.

    The transform (:class:`~.duration.Transform`) determines the actual length
    of Durable objects, and the properties (:class:`~.datatypes.Properties`)
    are forwarded to child contexts, where inside the
    :meth:`~.dom.lily.Music.time_length` of Music nodes values can be read and
    also modified.

    """
    def __init__(self, time, transform=None, properties=None):
        self.time = time    #: The :class:`Time` object we originate from.
        self.transform = transform or duration.Transform()      #: The current Transform.
        self.properties = properties or datatypes.Properties()  #: The current Properties.

    def __repr__(self):
        return "<{} time={} transform={} properties={}>".format(
            type(self).__name__, self.time, self.transform, self.properties)

    def _follow_trail(self, node, trail):
        """Compute length; return context, node and length at end of trail."""
        context = self
        length = 0
        for index in trail:
            context = context.enter(node)
            if index and node.is_sequential():
                length += node.time_length(context, index)
            node = node[index]
        return context, node, length

    def enter(self, node, time=None):
        """Return a new TimeContext.

        The returned TimeContext uses the new :class:`Time` (if given,
        otherwise the same as ours) and adds the :class:`~.duration.Transform`
        and the :class:`~.datatypes.Properties` of the specified ``node`` to
        the current ones.

        """
        t = self.transform
        transform = node.transform()
        if transform:
            t += transform
        p = self.properties
        properties = node.properties()
        if properties:
            p += properties
        return type(self)(time or self.time, t, p)

    def length(self, node, end=None):
        """Return the length of any node.

        Follows variable references using the scope (if given to the
        :class:`Time` instance) and calls :meth:`~.dom.lily.Music.time_length`
        for Music nodes. Returns 0 for any other, non-musical, node.

        """
        if isinstance(node, lily.Reference):
            return self.remote_length(node)
        elif isinstance(node, lily.Music):
            context = self.enter(node)
            return node.time_length(context, end)
        return 0

    def durable_length(self, node):
        """Return the length of a Durable node."""
        return self.transform.length(*self.time.get_duration(node))

    def remote_length(self, node):
        """Return the length of the value of an IdentifierRef node.

        Returns 0 if the node can't be found or is no music.

        """
        node, scope = node.get_value_with_scope(self.time.scope, self.time.wait)
        while isinstance(node, lily.Reference):
            node, scope = node.get_value_with_scope(scope, self.time.wait)
        if isinstance(node, lily.Music):
            time = type(self.time)(scope, self.time.wait)
            return node.time_length(self.enter(node, time), None)
        return 0

